prov/psm2/include/fi_ext_psm2.h

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2020 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_EXT_PSM2_H
+#define FI_EXT_PSM2_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define FI_PSM2_DOMAIN_OPS_1 "domain ops 1"
```

Comment:
At the very least, can you add psm2 to this string?  It turns out that GNI uses this exact same string to return a similar set of ops.  (GNI should have used a less generic string, but it's a little late for that now.)

The ops are generic enough to be used by any object and any provider.  Maybe it would be better to extend fi_control() with this behavior.  Something like:
```
struct fi_fid_var {
    int var;
    void *val;
};

static inline int fi_set_val(struct fid *fid, int var, void *val)
{
    struct fi_fid_var var;
    var.var = var;
    var.val = val;
    return fi_control(fid, FI_SET_VAL, &var);
}
```
With a similar wrapper for fi_get_val().  Then we can handle var id's on a case by case basis, to see if it can be defined as a common id, or should be provider specific.

As for disconnect, if I'm understanding what it does, when set, the disconnect protocol is disabled.  I'm at a toss up whether this would be useful as a generic setting or psm2 specific setting.  Associating it with the domain, rather than the endpoint seems more of a psm2 limitation.

Reply:
The disconnect protocol is a very psm2 specific thing. It applies to an object corresponding hw context, which doesn't have 1:1 mapping to endpoint. Setting it at the domain level is the easiest path to achieve the intended goal -- to solve an issue with applications that are linked to middle-ware libraries that have opposite requirement of the disconnection protocol. 

General variable setting via fi_control is a good idea. However, we really don't have a set of common vars to start with. In addition, fi_open_ops() is the official API for provider specific extensions, do we want to extend part of the functionality into fi_control?

I will fix the string first, and leave the other parts open to debate.

Reply:
Both psm2 and gni are using open_ops to expose get_val/set_val functions.  Rather than have multiple providers expose similar functionality through open_ops, I would rather move to a common solution.  Since get_val/set_val are basically designed to be called once by the application, fi_control seems more reasonable than defining a common open_ops structure.  I'm okay with carving out provider specific identifiers there.

As an alternative, we could expose a new API, fi_setparams(), which could be used as a peer to fi_getparams().  That could provide a programmatic way to set parameters that may not be associated with a fid.  This would likely require additional work in order to plumb the call to the providers.  My preference is the easier path. :)

Reply:
ok, let me work out a prototype for the fi_control path.

Reply:
See #6386